cmd/gardenlet/app/app.go

```diff
@@ -508,6 +515,41 @@ func removeLegacyShootStateControllerFinalizerFromSecrets(ctx context.Context, s
 	return flow.Parallel(taskFns...)(ctx)
 }
 
+func (g *garden) cleanupStaleShootStates(ctx context.Context, gardenClient client.Client) error {
+	if err := gardenClient.Get(ctx, client.ObjectKey{Name: g.config.SeedConfig.Name, Namespace: v1beta1constants.GardenNamespace}, &seedmanagementv1alpha1.ManagedSeed{}); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return fmt.Errorf("failed checking whether gardenlet is responsible for a managed seed: %w", err)
+		}
+		return nil
+	}
+
+	g.mgr.GetLogger().Info("Removing stale ShootState resources from garden cluster since I'm responsible for a managed seed (GEP-22)")
+
+	shootList := &gardencorev1beta1.ShootList{}
+	if err := gardenClient.List(ctx, shootList, client.MatchingFields{gardencore.ShootSeedName: g.config.SeedConfig.Name}); err != nil {
+		return err
+	}
+
+	var taskFns []flow.TaskFn
+
+	for _, s := range shootList.Items {
+		shoot := s
+
+		// If status.seedName is different than seed name gardenlet is responsible for, then a migration takes place.
+		// In this case, we don't want to delete the shoot state. It will be deleted eventually after successful
+		// restoration.
+		if shoot.Status.SeedName != nil && *shoot.Status.SeedName != g.config.SeedConfig.Name {
```

Comment:
In the restore phase, `spec.seedName` will be equal to `status.seedName`, but we shouldn't delete the `ShootState` as we might still need it for restoration. Correct?

Reply:
True, good point. I've added another condition, PTAL :)